docs/okeps/okep-5233-pre-assigned-pod-net-id.md

```diff
@@ -0,0 +1,344 @@
+    - expression: "('v1.multus-cni.io/default-network' in oldObject.metadata.annotations) == ('v1.multus-cni.io/default-network' in object.metadata.annotations)"
+      message: "The 'v1.multus-cni.io/default-network' annotation cannot be changed after the pod was created"
+```
+The `NetworkSelectionElement` structure has an extensive list of fields, this enhancement 
+focuses only on the following:
+```cgo
+type NetworkSelectionElement struct {
+	// IPRequest contains an optional requested IP addresses for this network
+	// attachment
+	IPRequest []string `json:"ips,omitempty"`
+	// MacRequest contains an optional requested MAC address for this
+	// network attachment
+	MacRequest string `json:"mac,omitempty"`
+	// IPAMClaimReference container the IPAMClaim name where the IPs for this
+	// attachment will be located.
+	IPAMClaimReference string `json:"ipam-claim-reference,omitempty"`
+}
+```
+
+Any other field set in the struct will be ignored by OVN-Kubernetes.
+Both `Name` and `Namespace` are ignored as the primary UDN is known and cannot be changed.
+Note that `GatewayRequest` is not listed either, the default gateway is an attribute of the network is not going to be
+configurable per pod.
+With `k8s.ovn.org/primary-udn-ipamclaim` being deprecated in favor of the `IPAMClaimReference` field
+in the `NetworkSelectionElement` we have to define the expected behavior. To avoid conflicting 
+settings when `v1.multus-cni.io/default-network` is set the `k8s.ovn.org/primary-udn-ipamclaim` is 
+going to be ignored, it will be reflected in the opposite scenario for backwards compatibility 
+with a plan to remove it in a future release.
+Deprecation plan for the `k8s.ovn.org/primary-udn-ipamclaim` annotation:
+- release-N - emit a warning event stating that the annotation is deprecated and will be removed in a future release.
+- release-N+1 - fail to configure pods with the annotation set.
+- release-N+2 - remove any code handling the annotation, effectively ignoring it.
+
+### Address allocation
+
+OVN-Kubernetes currently [generates](https://github.com/ovn-kubernetes/ovn-kubernetes/blob/3ef29b9a32b04b7917a0afd6b0e9651d17242ed7/go-controller/pkg/util/net.go#L100-L113)
+the overlay MAC addresses from the IPs:
+
+- IPv4: It takes the four octets of the address (e.g `AA.BB.CC.DD`) and uses them to
+create the MAC address with a constant prefix (e.g. `0A:58:AA:BB:CC:DD`).
+- IPv6: Computes a SHA256 checksum from the IPv6 string and uses the first four bytes for the MAC
+address with the `0A:58` constant prefix(e.g. `0A:58:SHA[0]:SHA[1]:SHA[2]:SHA[3]`).
+
+Although unlikely, we need to implement logic that ensures that the MAC address requested through
+the `NetworkSelectionElement` does not conflict with any other configured address on the UDN
+(including addresses consumed by OVN-Kubernetes).
+A similar approach is required for IP address conflict detection.
+When a conflict is detected the pod should not start and an appropriate event should be emited.
```

Comment:
Do we know if OVN is already checking that ?

Reply:
I'm afraid **not**, but we need to check. 

Can we have an AP for checking this out ? 
@ormergi could you check it out if you don't mind ? Feel free to create a JIRA issue to track this PoC, and ensure that's listed in our board. Please reach out to me for this.

Reply:
> Do we know if OVN is already checking that ?

If OVN doesn't hard fail on this predictably we are going to have to check it from ovn-k.

Reply:
According to the core OVN maintainers, duplicate IP/MACs are misconfigs.

Quoting them: 
> What happens should probably qualify as "undefined behavior".

Meaning, we'll need to implement this logic on OVN-Kubernetes.